src/audio/smart_amp/smart_amp_maxim_dsm.c

```diff
@@ -0,0 +1,459 @@
+//
+// Copyright(c) 2020 Maxim Integrated All rights reserved.
+//
+// Author: Ryan Lee <ryans.lee@maximintegrated.com>
+
+#include <sof/audio/component.h>
+#include <sof/trace/trace.h>
+#include <sof/drivers/ipc.h>
+#include <sof/ut.h>
+
+#include <sof/audio/channel_map.h>
+#include <sof/trace/trace.h>
+#include <user/trace.h>
+#include <sof/bit.h>
+#include <sof/common.h>
+#include <ipc/channel_map.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <sof/audio/smart_amp/smart_amp.h>
+
+static int maxim_dsm_init(struct smart_amp_mod_struct_t *hspk, struct comp_dev *dev)
+{
+	/* Maxim Dynamic Speaker Manangement initialization */
+	return 0;
+}
+
+static inline void *_memmove(void *dest, const void *src, size_t count)
+{
+	char *tmp;
+	const char *s;
+
+	if (dest <= src) {
+		tmp = dest;
+		s = src;
+		while (count--)
+			*tmp++ = *s++;
+	} else {
+		tmp = dest;
+		tmp += count;
+		s = src;
+		s += count;
+		while (count--)
+			*--tmp = *--s;
+	}
+
+	return dest;
+}
+
+static void maxim_dsm_ff_proc(struct smart_amp_mod_struct_t *hspk,
+			      struct comp_dev *dev, void *in, void *out,
+			      int nsamples, int szsample)
+{
+	union smart_amp_buf buf, buf_out;
+	int16_t *input = (int16_t *)hspk->buf.input;
+	int16_t *output = (int16_t *)hspk->buf.output;
+	int *w_ptr = &hspk->buf.ff.avail;
+	int *r_ptr = &hspk->buf.ff_out.avail;
+	bool is_16bit = szsample == 2 ? true : false;
+	int remain;
+	int x;
+
+	buf.buf16 = (int16_t *)hspk->buf.ff.buf;
+	buf.buf32 = (int32_t *)hspk->buf.ff.buf;
+	buf_out.buf16 = (int16_t *)hspk->buf.ff_out.buf;
+	buf_out.buf32 = (int32_t *)hspk->buf.ff_out.buf;
+
+	if (*w_ptr + nsamples <= DSM_FF_BUF_DB_SZ) {
+		if (is_16bit)
+			memcpy_s(&buf.buf16[*w_ptr], nsamples * szsample,
+				 in, nsamples * szsample);
+		else
+			memcpy_s(&buf.buf32[*w_ptr], nsamples * szsample,
+				 in, nsamples * szsample);
+		*w_ptr += nsamples;
+	} else {
+		comp_err(dev,
+			 "[DSM] Feed Forward buffer overflow. (w_ptr : %d + %d > %d)",
+			 *w_ptr, nsamples, DSM_FF_BUF_DB_SZ);
+		return;
+	}
+
+	if (*w_ptr >= DSM_FF_BUF_SZ) {
+		/* Do FF processing */
+		for (x = 0; x < DSM_FRM_SZ; x++) {
+			/* Buffer ordering for DSM : LRLR... -> LL...RR... */
+			if (is_16bit) {
+				input[x] = (buf.buf16[2 * x]);
+				input[x + DSM_FRM_SZ] =
+					(buf.buf16[2 * x + 1]);
+			} else {
+				input[x] = (buf.buf32[2 * x] >> 16);
+				input[x + DSM_FRM_SZ] =
+					(buf.buf32[2 * x + 1] >> 16);
+			}
+		}
+
+		remain = (*w_ptr - DSM_FF_BUF_SZ);
+		if (remain) {
+			if (is_16bit)
+				_memmove(&buf.buf16[0],
+					 &buf.buf16[DSM_FF_BUF_SZ],
+					 remain * szsample);
+			else
+				_memmove(&buf.buf32[0],
+					 &buf.buf32[DSM_FF_BUF_SZ],
+					 remain * szsample);
+		}
+		*w_ptr -= DSM_FF_BUF_SZ;
+
+		/* Do DSM feed forward processing here. Currently bypassed */
+		memcpy_s(output, DSM_FF_BUF_SZ * szsample,
+			 input, DSM_FF_BUF_SZ * szsample);
+
+		for (x = 0; x < DSM_FRM_SZ; x++) {
+			/* Buffer re-ordering LR/LR/LR */
+			if (is_16bit) {
```

Comment:
Do the type check outside of the for loop.

Reply:
Agreed. Shall fix it.